### PR TITLE
Implement `use_standalone()`

### LINKED ISSUE #1654: Provide `use_compat()` for managing common `compat-[pkg].R` files
Often when writing a new package I find myself simply lifting scripts like [`compat-purrr.R`](https://github.com/r-lib/rlang/blob/main/R/compat-purrr.R) from the _{rlang}_ source-code to avoid taking on more dependencies than necessary. I think formalising this process would be well worth the effort. E.g. it could be done as follows:

* `use_compat(pkg)` would copy a `compat-pkg.R` into the user's `R` directory, e.g. from the github for an existing package such as _{rlang}_ where the 'master' version is defined, or from a dedicated repo for tidyverse `compat-` files. The second option seems better to me as it would also provide a nice place to look for such files without necessarily having to go through _{usethis}_.
* Re-running `use_compat()` after the 'master' version has changed could update the existing `compat-` script after prompting the user to confirm any (potentially breaking) changes.
* `use_compat()` could also support single functions, e.g. in the case of `withr::defer()`, which has its own `compat-defer.R` file in the _{rlang}_ source code.
* Specifically, I think standardising 'compat' files for at least the following packages would be really useful: _{cli}_, _{dplyr}_, _{forcats}_, _{glue}_, _{purrr}_, _{rlang}_, _{stringr}_, _{vctrs}_, _{withr}_ and _{zeallot}_

Obviously the real value here would be in having such `compat-` scripts readily available. However it sees like this is already somewhat of a 'soft-supported' service from the Tidyverse, to wit, [`rlang/compat-vctrs.R`](https://github.com/r-lib/rlang/blob/main/R/compat-vctrs.R) begins with the following line:
```
# Latest version: https://github.com/r-lib/rlang/blob/main/R/compat-vctrs.R
```
So, perhaps the work is already largely done and the creation of `use_compat()` would mainly be a case of compiling all `compat-` files which already exist.

### PR DESCRIPTION
**Pull request by hadley**
Fixes #1654

### CODE REVIEW

File: R/use_github_file.R [hunk 1]
```diff
@@ -50,41 +50,31 @@ use_github_file <- function(repo_spec,
...
-  res <- gh::gh(
-    "/repos/{repo_spec}/contents/{path}",
-    repo_spec = repo_spec, path = path,
+  lines <- read_github_file(
+    repo_spec = repo_spec,
+    path = path,
     ref = ref,
-    .destfile = tf,
-    .accept = "application/vnd.github.v3.raw"
+    host = host
```
Comment (hadley) [inline R/use_github_file.R:75]: I think `host` previously had no effect?

**Comment by hadley** [comment]
Requires https://github.com/r-lib/rlang/pull/1548

File: NEWS.md [hunk 2]
```diff
@@ -17,6 +20,7 @@
 * `use_travis()` and `use_appveyor()` are now defunct because we no longer  
   recommend Travis or Appveyor. We recommend GitHub actions instead (#1517).
 
+>>>>>>> 8a4a633e6490d5be70734c03844987fb58a2904d
```
Comment (jennybc) [inline NEWS.md:23]: 👀

File: R/tidyverse.R [hunk 3]
```diff
@@ -132,11 +132,7 @@ use_tidy_dependencies <- function() {
   # If needed, copy in lightweight purrr compatibility layer
   if (!proj_desc()$has_dep("purrr")) {
     use_directory("R")
-    use_github_file(
-      "r-lib/rlang",
-      path = "R/compat-purrr.R",
-      save_as = "R/compat-purrr.R"
-    )
+    use_compat("r-lib/rlang", "purrr")
```
Comment (jennybc) [inline R/tidyverse.R:135]:
```suggestion
    use_standalone("r-lib/rlang", "purrr")
```

File: R/use-standalone.R [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+#' that it can be copied into another package. `use_standalone()` makes it easy
+#' to get such a file into your own repo.
+#'
+#' @inheritParams create_from_github
+#' @param file Name of standalone file. The `standalone-` prefix and file
+#'   extension are optional. If omitted, will allow you to choose from the
+#'   standalone files offered by that repo.
+#' @export
+use_standalone <- function(repo_spec, file = NULL) {
+  proj_get() # force project discovery
```
Comment (jennybc) [inline R/use-standalone.R:13]: `check_is_project()`?

File: R/use-standalone.R [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+  if (is.null(file)) {
+    file <- standalone_choose()
```
Comment (jennybc) [inline R/use-standalone.R:16]: Should we pass `repo_spec` here?

File: R/use-standalone.R [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    cli::cli_abort(
+      "No standalone files found in {repo_spec}.",
+      call = error_call
+    )
+  }
+
+  if (!is_interactive()) {
+    cli::cli_abort(
+      c(
+        "`file` is absent, but must be suppled.",
```
Comment (jennybc) [inline R/use-standalone.R:61]:
```suggestion
        "`file` is absent, but must be supplied.",
```

File: tests/testthat/test-use-standalone.R [hunk 7]
```diff
@@ -0,0 +1,39 @@
+test_that("can import standalone file with dependencies", {
```
Comment (jennybc) [inline tests/testthat/test-use-standalone.R:1]: `skip_if_offline()`?

**Review by jennybc — COMMENTED** [review]
I realize this is a draft and is unfinished, but I made one pass on my way through the usethis notifications.

**Comment by jennybc** [comment]
Note to self: unskip "use_tidy_dependencies() isn't overly informative" test when this concludes.

File: R/use-standalone.R (hunk 5 above)
Comment (hadley) [reply R/use-standalone.R:16]: Oops, yes.

File: tests/testthat/test-use-standalone.R (hunk 7 above)
Comment (hadley) [reply tests/testthat/test-use-standalone.R:1]: I used `skip_on_cran()` to be consistent with the other test. But should I switch both to use `skip_if_offline()`?

File: R/use-standalone.R [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      file <- paste0("standalone-", file)
+    }
+  }
+
+  src_path <- path("R", file)
+  dest_path <- path("R", paste0("import-", file))
+
+  lines <- read_github_file(repo_spec, path = src_path)
+  lines <- c(standalone_header(repo_spec, src_path), lines)
+  write_over(proj_path(dest_path), lines, overwrite = TRUE)
```
Comment (DavisVaughan) [inline R/use-standalone.R:31]: Should it be documented that it always overwrites an existing standalone file?

File: R/use-standalone.R [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+  dependencies <- standalone_dependencies(lines, path)
+  for (dependency in dependencies) {
+    use_standalone(repo_spec, dependency)
+  }
+
+  invisible()
+}
+
+standalone_choose <- function(repo_spec, error_call = parent.frame()) {
```
Comment (DavisVaughan) [inline R/use-standalone.R:41]:
```suggestion
standalone_choose <- function(repo_spec, error_call = caller_env()) {
```

File: R/use-standalone.R [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+standalone_header <- function(repo_spec, path) {
+  c(
+    "# Standalone file: do not edit by hand",
+    glue("# Source: <https://github.com/{repo_spec}/blob/main/{path}>"),
+    paste0("# ", strrep("-", 72 - 2)),
+    "#"
+  )
+}
+
+standalone_dependencies <- function(lines, path, call = caller_env()) {
```
Comment (DavisVaughan) [inline R/use-standalone.R:89]:
```suggestion
standalone_dependencies <- function(lines, path, error_call = caller_env()) {
```

Just to be consistent with `standalone_choose()`?

File: R/use-standalone.R [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        "`file` is absent, but must be supplied.",
+        i = "Possible options are {.or {choices}}."
+      ),
+      call = error_call
+    )
+  }
+
+  choice <- utils::menu(
+    choices = choices,
+    title = "Which standalone file do you want to use (0 to exit)"
```
Comment (DavisVaughan) [inline R/use-standalone.R:71]:
```suggestion
    title = "Which standalone file do you want to use (0 to exit)?"
```

File: R/use-standalone.R [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  )
+  if (choice == 0) {
+    cli::cli_abort("Selection cancelled")
```
Comment (DavisVaughan) [inline R/use-standalone.R:74]: `error_call`?

File: R/use-standalone.R [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  names[choice]
```
Comment (DavisVaughan) [inline R/use-standalone.R:77]:
```suggestion
  names[[choice]]
```

Just to make it clear that there will only be 1

**Review by DavisVaughan — APPROVED** [review]
I used it here and it worked well! https://github.com/tidyverse/dplyr/pull/6722

File: R/use-standalone.R [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+#'
+#' @description
+#' A "standalone" file implements a minimum set of functionality in such a way
+#' that it can be copied into another package. `use_standalone()` makes it easy
+#' to get such a file into your own repo.
+#'
+#' It always overwrites an existing standalone file of the same name, making
+#' it easy to update previously imported code.
+#'
+#' @inheritParams create_from_github
```
Comment (jennybc) [inline R/use-standalone.R:11]: This makes the following promise about `repo_spec`:

> `repo_spec`
> A string identifying the GitHub repo in one of these forms:
>
> Plain OWNER/REPO spec
>
> Browser URL, such as "https://github.com/OWNER/REPO"
>
> HTTPS Git URL, such as "https://github.com/OWNER/REPO.git"
>
> SSH Git URL, such as "git@github.com:OWNER/REPO.git"
>
> In the case of a browser, HTTPS, or SSH URL, the host is extracted from the URL. The REPO part will be the name of the new local folder, which is also a project and Git repo.

But no such parsing code is here.

Which also brings up the bigger issue which is that this function doesn't support GHE. In general, our stance is to make stuff like this work for github.com and GHE. I think the toolkit we use elsewhere works well here. I can do it if you want.

File: tests/testthat/test-use-standalone.R (hunk 7 above)
Comment (jennybc) [reply tests/testthat/test-use-standalone.R:1]: I suspect the other test probably first appeared in the era before `skip_if_offline()` implied `skip_on_cran()`. So yeah I think `skip_if_offline()` is better for both now.

**Review by jennybc — COMMENTED** [review]
I think the main thing to settle is whether to make this work for GHE. I think we should do it, since we've met that standard everywhere else.

File: tests/testthat/test-use-standalone.R (hunk 7 above)
Comment (jennybc) [reply tests/testthat/test-use-standalone.R:1]: I just did this.

File: R/use-standalone.R (hunk 14 above)
Comment (hadley) [reply R/use-standalone.R:11]: I switched to `create_from_github()` `repo_spec` and actually parsed it, and added a `host` argument. LMK if there's anything else needed for GHE support.

File: R/use-standalone.R [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+  check_is_project()
+
+  parsed_repo_spec <- parse_repo_url(repo_spec)
+  if (!is.null(parsed_repo_spec$host)) {
+    repo_spec <- parsed_repo_spec$repo_spec
+    host <- parsed_repo_spec$host
+  }
+
+  if (is.null(file)) {
+    file <- standalone_choose(repo_spec)
```
Comment (jennybc) [inline R/use-standalone.R:27]:
```suggestion
    file <- standalone_choose(repo_spec, host = host)
```

**Review by jennybc — APPROVED** [review]
With the caveat that I didn't manually test on GHE, yes, I think this is good to go (after you take my suggestion to pass `host` to `standalone_choose()`).

One last suggestion: in `create_from_github()`, perhaps we should clarify the docs a bit, now they are being inherited elsewhere.

Such as:

**For `create_from_github()`**, the `REPO` part will be the name of the new local folder, which is also a project and Git repo.

**Comment by hadley** [comment]
@jennybc the docs were a bit tangled, but I think I've made an improvement. Anyway, I'm going to merge this PR so it's off my plate, and you should feel free to tweak the docs as you see fit.

**Pull request merged**